Sync episode actions with gpodder, smart mark as played

mfietz opened the pull request:
Closes AntennaPod/AntennaPod#539
Closes AntennaPod/AntennaPod#542
Closes AntennaPod/AntennaPod#694

Episode actions are recorded and synced when user is logged into gpodder.

* Create episode actions when episodes are downloaded, played, deleted and marked as read
* Sync (download and upload) episode actions
* MediaPlayerActivity deletes almost completely played episode on close
* Improved parsing of datetime strings
* Smart mark as played can be disabled or set in the preferences

![Screenshot](https://cloud.githubusercontent.com/assets/6860662/6952930/3abeb508-d8c6-11e4-9eaf-aeb4c69f92c8.png)

## For debugging
You can goto https://gpodder.net/share/ and click the world icon near User page. Recently played doesn't seem to update immediately... Maybe the web server is caching the request

What I usually do:
$ curl -i -u [Username]:[Password] -H "Content-Type: application/json" -H "Accept: application/json" -X GET https://gpodder.net/api/2/episodes/[Username].json?since=[Timestamp]

Timestamp is seconds since 1.1.1970, set it to 0 or get the current time from http://www.epochconverter.com/ (For debugging, I take the current time and set the last three digits to zero).
You should see some nice JSON ;)

You should also see the syncing in logcat

mfietz commented:
Please do not merge yet. There will be another commit on the branch later today.

Problem: After fetching actions from the server, we should dismiss all episode play actions (set position) if there is a newer play action for this episode in the queue.

mfietz commented:
That should do it...

TomHennen commented:
Yuck, it looks like that last PR I merged to develop broke the ability to auto-merge this one.  Could you take a look at it?

Also, I tested some of this out in an emulator. It looked good.  I haven't really been able to see the multi-device syncing yet, that will probably take more testing. I'm considering just pushing it to the alpha channel and seeing what people think.

mfietz commented:
Should be working again.
Btw, you may have noticed that I remove "if (BuildConfig.DEBUG)" whenever I see it. Doesn't seem to work with Android Studio anymore.
It should be rather easy to make ProGuard remove any Log.* statements for the release APK.

TomHennen commented:
Just noticed one thing when testing this in a 2.3.7 and 5.0 emulator, if I play the episode to the point at which it would be considered 'played' and then reset the playback position to some earlier time, then the item is gone from the queue.  It appears the user would need to go back to All Episodes (or Playback History) and explicitly add the item again.  Does this behavior make sense?

I'm not really sure myself. Putting it back in the queue after it's been removed sounds wrong.

I almost wonder if the episode is being removed from the queue too early. That we should wait to remove it until the user somehow navigates away from the playback activity (or android puts us to sleep).  What do you think?

mfietz commented:
I guess you are right. At the moment, the episode is marked as read/played, removed from the queue and added to the history as soon as the user pauses it after the threshold. This happens in the playback service.

The file itself is only deleted if the user closes the audio player activity (or she could not press play again). Moving all the smart played logic into the activity would solve removing the episode too early. Only the last playback position would count.

But I have to investigate if this logic is needed elsewhere, e.g. in some widget.

TomHennen commented:
Ok. I'll hold off on this PR for now then.

Working on getting log messages removed from release builds now.

mfietz commented:
http://proguard.sourceforge.net/manual/examples.html#logging

TomHennen commented:
Yup, definitely easy. Should get it pushed today.

mfietz commented:
Hm, say you are on the queue fragment, pause and smart mark as played fires.
What should happen? Should the item be removed from the queue (might be odd), should the media be deleted?
Should all played items be removed and their media deleted if the queue fragment is closed?
Same goes for the new episode fragment

TomHennen commented:
This is confusing.

One thing I don't quite understand is why people want the media deleted
immediately. For me it's enough that it gets deleted when AntennaPod starts
downloading more things. Are people space constrained and the need the
space back ASAP?  If we didn't delete right away would that make things
more rational? (just as a thought experiment)

I don't know how feasible this is, but perhaps the best indicator that the
user is done with an episode (when using smart delete) is when they start
playing *something else*.

If we go with that policy then navigating away, having the app closed, etc
would have no effect. However if you are in the smart window and you start
playing another podcast *then* it gets removed and the media deleted.  The
more I think about it the more I like this idea. Seems intuitive. Would it
be feasible?

On Fri, Apr 3, 2015, 1:45 PM Martin Fietz <notifications@github.com> wrote:

> Hm, say you are on the queue fragment, pause and smart mark as played
> fires.
> What should happen? Should the item be removed from the queue (might be
> odd), should the media be deleted?
> Should all played items be removed and their media deleted if the queue
> fragment is closed?
> Same goes for the new episode fragment
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AntennaPod/AntennaPod/pull/706#issuecomment-89413919>.
>

mfietz commented:
This is actually the way I already have implemented it. I should probably write some tests... 

Auto delete: I get why some people like it. My smartphone has 16 GB space. With MP3, about 16h fill up 1 GB. I have several podcasts where an episode is several hours long.
Android should also be able to play FLAC - one episode could already be 1GB.

It is probably best to show at least the undo bar on auto removing the episode from the queue. Episodes disappearing would be rather confusing.
But I have to think about how to "un-delete" or rather delete the file later if it is safe. Perhaps I can listen for the undobar to disappear. I would delete the media after the user missed her opportunity to undo smart mark as played (for the lolz, henceforward this shall be called "to be smayed" ^^)

TomHennen commented:
Hmm, it seemed like it had been removing from the queue as soon as playback
was paused and didn't seem to care if you actually played another episode.
The two behaviors are different in my mind.

I like the idea of undo and listening for undo to disappear. That seems
like a good idea no matter what as it lets the user know what's going on.

SMAYED will definitely be a useful feature. :-)

On Fri, Apr 3, 2015, 2:11 PM Martin Fietz <notifications@github.com> wrote:

> This is actually the way I already have implemented it. I should probably
> write some tests...
>
> Auto delete: I get why some people like it. My smartphone has 16 GB space.
> With MP3, about 16h fill up 1 GB. I have several podcasts where an episode
> is several hours long.
> Android should also be able to play FLAC - one episode could already be
> 1GB.
>
> It is probably best to show at least the undo bar on auto removing the
> episode from the queue. Episodes disappearing would be rather confusing.
> But I have to think about how to "un-delete" or rather delete the file
> later if it is safe. Perhaps I can listen for the undobar to disappear. I
> would delete the media after the user missed her opportunity to undo smart
> mark as played (for the lolz, henceforward this shall be called "to be
> smayed" ^^)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AntennaPod/AntennaPod/pull/706#issuecomment-89421376>.
>

mfietz commented:
Sorry, I meant implemented *locally*. Haven't pushed the new behavior yet.

TomHennen commented:
Ah, gotcha. That's perfect.

On Fri, Apr 3, 2015, 2:24 PM Martin Fietz <notifications@github.com> wrote:

> Sorry, I meant implemented *locally*. Haven't pushed the new behavior yet.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AntennaPod/AntennaPod/pull/706#issuecomment-89424291>.
>

mfietz commented:
How do you feel about greenrobot's EventBus* replacing EventDistributor? I uses it in another project and found it quite useful.

I have the following problem: When smark on the queue fragment, the corresponding feedmedia is marked as played. But the EventDistributor only informs the Queue to reload its items, not which item actually was removed. I need to know which media was marked to show the undobar for it.
With EventBus we would not send simple event codes as EventDistributor does now, but we could send additional information (wrapped in a custom class).
I also think the whole playing and pausing could get a lot easier. Instead of sending an intent through several callbacks and classes, we would simple send an Event (from any activity/fragment we want) that the playback service can receive directly.

Replacing the whole system at once would be quite a lot of work. For now, I would only send an additional Event after the Distributor sent its code. Perhaps I would even replace the whole "QUEUE_UPDATE" code with new logic.
The remaining Distributor code would get replaced whenever there is a need (or there is nothing else to do *g*).

--
* https://github.com/greenrobot/EventBus

TomHennen commented:
I'm honestly not familiar with EventBus.

I don't have a strong opinion either way. If you think it will be helpful
in the long run, and we do it gradually (as you suggest) then it could be a
good idea.

On Sat, Apr 4, 2015, 1:33 PM Martin Fietz <notifications@github.com> wrote:

> How do you feel about greenrobot's EventBus* replacing EventDistributor? I
> uses it in another project and found it quite useful.
>
> I have the following problem: When smark on the queue fragment, the
> corresponding feedmedia is marked as played. But the EventDistributor only
> informs the Queue to reload its items, not which item actually was removed.
> I need to know which media was marked to show the undobar for it.
> With EventBus we would not send simple event codes as EventDistributor
> does now, but we could send additional information (wrapped in a custom
> class).
> I also think the whole playing and pausing could get a lot easier. Instead
> of sending an intent through several callbacks and classes, we would simple
> send an Event (from any activity/fragment we want) that the playback
> service can receive directly.
>
> Replacing the whole system at once would be quite a lot of work. For now,
> I would only send an additional Event after the Distributor sent its code.
> Perhaps I would even replace the whole "QUEUE_UPDATE" code with new logic.
> The remaining Distributor code would get replaced whenever there is a need
> (or there is nothing else to do *g*).
>
>    - https://github.com/greenrobot/EventBus
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AntennaPod/AntennaPod/pull/706#issuecomment-89613956>.
>

mfietz commented:
Thank god I don't have to review this... ^^

TomHennen commented:
:(

mfietz commented:
Btw, I'm not sure why Android Studio keeps reorganizing the imports. Maybe we should agree on a common Java Code Style and have a "reformat the whole project" commit. Would fix those damn tabs...

Would also make it quite difficult to compare pre-reformat and after-reformat commits...

TomHennen commented:
I know. I always *want* to do something like that, but I'm very scared of the effective loss of history.  I wouldn't worry about it for now.

We could agree on a style and just use that for all new things. If possible we could/should commit the style choices so that AntennaPod uses them for everyone (or at least makes it easy to apply them).

mfietz commented:
I was wondering, should episodes swiped from the queue or new episodes after they have been marked as read also be added to the playback history?
(Why do we even have this? As useless as Downloads (or rather the Dowload Log), if you ask me)

TomHennen commented:
I think they should only be added to playback history if they've been played. Removing them from either list doesn't really count.

I've been using playback history while testing this feature, but I never use it at any other time. Not sure if anyone else does.

Regarding this particular PR. I thought that we had discussed not removing the episode from the queue until a user plays the next episode if SMAYED. The behavior that I'm currently seeing is the old behavior. Where the episode is removed as soon as it enters the SMAYED zone. I also don't see the undo feature that was discussed. Were you thinking of doing those things later?

TomHennen commented:
btw, I put a debug build of this PR here: https://www.dropbox.com/s/uvbeh285jtpd3lu/AntennaPod-debug-PR-706.apk?dl=0

It would be awesome if people could test it out.

mfietz commented:
What I think I have implemented...
When the user is on the Queue fragment and an episode is playing, goes over the threshold and then plays another item, smark fires. You should see an undobar (unless I fucked up)
It the user is playing an episode in the audio player activity (the one with the big cover), pauses and then pushes return, smark also fires.

TomHennen commented:
Ah, I was doing it the second way you described.

Why do you think the audio player should have different behavior?  I suppose you could argue it's a very deliberate action on the part of the user.  Shouldn't we show the undo bar there as well though?

mfietz commented:
The stuff happening when leaving the audio player activity is probably completely unnecessary.
Showing an undobar would be rather tricky, the stuff happens in the audio player activity, the queue fragment is not receiving any actions (from both EventDistributor and EventBus) at this point

TomHennen commented:
Ok, I see.

FYI, this force closes if you try to disable it.

```
04-05 17:38:02.619: E/AndroidRuntime(1651): FATAL EXCEPTION: main
04-05 17:38:02.619: E/AndroidRuntime(1651): Process: de.danoeh.antennapod.debug, PID: 1651
04-05 17:38:02.619: E/AndroidRuntime(1651): java.lang.RuntimeException: Unable to create application de.danoeh.antennapod.PodcastApp: java.lang.NumberFormatException: Invalid int: "off"
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4521)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at android.app.ActivityThread.access$1500(ActivityThread.java:144)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1339)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at android.os.Handler.dispatchMessage(Handler.java:102)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at android.os.Looper.loop(Looper.java:135)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at android.app.ActivityThread.main(ActivityThread.java:5221)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at java.lang.reflect.Method.invoke(Native Method)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at java.lang.reflect.Method.invoke(Method.java:372)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
04-05 17:38:02.619: E/AndroidRuntime(1651): Caused by: java.lang.NumberFormatException: Invalid int: "off"
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at java.lang.Integer.invalidInt(Integer.java:138)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at java.lang.Integer.parse(Integer.java:410)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at java.lang.Integer.parseInt(Integer.java:367)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at java.lang.Integer.parseInt(Integer.java:334)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at java.lang.Integer.valueOf(Integer.java:525)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at de.danoeh.antennapod.core.preferences.UserPreferences.loadPreferences(UserPreferences.java:142)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at de.danoeh.antennapod.core.preferences.UserPreferences.<init>(UserPreferences.java:103)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at de.danoeh.antennapod.core.preferences.UserPreferences.createInstance(UserPreferences.java:116)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at de.danoeh.antennapod.PodcastApp.onCreate(PodcastApp.java:41)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1011)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4518)
04-05 17:38:02.619: E/AndroidRuntime(1651): 	... 9 more
```

mfietz commented:
Gosh, that bug was stupid...

TomHennen commented:
No problem.

Oh, and could you update about.html to include the EventBus library?

TomHennen commented:
I think the inclusion of EventBus might have messed up swipe to remove in the queue.

Swipe to remove still works, but you still see the undo bar saying 'item removed' if you drag an item up or down.

This is probably happening because DBWriter.moveQueueItemHelper is issuing a 'REMOVED' and 'ADDED' event when someone tries to move an item.  QueueFragment.onEventMainThread is then seeing the 'action == REMOVED' and displaying the undo bar.

TwoD commented:
I see this thread is currently discussing the playback history and whatnot, but If I get this right, the original message here says gPodder episode syncing has been implemented, no?

Well, nothing about that works at all. AntennaPod never syncs anything about episodes with gpodder.net. Issues about this keep getting closed, but I can't find any solution, or even answers to what's supposed to work.

I get many many many error messages about gpodder.net syncing a day, even when I haven't touched the app in days it's still bugging me about something going horribly wrong.

I'm frankly sick of it, mostly because I can't read any of the messages because the text doesn't fit in the notification area, and there's no logging of what it's trying to do or what went wrong.

I've checked my gpodder.net credentials over and over, nothing wrong there. Syncing episode states works just fine between my home and work desktop machines, but my phone never triggers any activity visible using the gpodder.net API or the web UI (which, I admit, is completely broken as half the pages just return "500 internal server error").

I'm sorry if I ranted in the wrong place. If so, please point me to the correct open issue about gpodder.net episode syncing.

mfietz commented:
@TwoD Please do not report issues in a pull request. (Pretty much ignored it when I first saw it).
When reporting an issue, please adhere to https://github.com/AntennaPod/AntennaPod/blob/master/CONTRIBUTING.md#how-to-report-a-bug

"Well, nothing about that works at all.", "I get many many many error messages about gpodder.net syncing" (Imagine some older relative calling, telling you that their computer isn't working...)
-> usefulness: 0

TwoD commented:
@mfietz Sorry, but I'm only bringing it up here because this appears to be the pull request in which episode syncing was added. I know there already is a bunch of issues about this problem, but I wanted to bring attention to the problems here as this is where the feature was added. (Though in the issues it looks as if people still think the feature doesn't exist yet.)

The error description is useless, because the error output/logging is useless. I can't read most of what it says, and the parts I can read (when it complains about wrong credentials) are obviously incorrect. "It never syncs and throws an unreadable error" is all the info I've got...
The unhelpful messages were reported in #569 but there's been no change there yet so I can't give a better error description even though I want to.

I tried looking through the gpodder code to see what could go wrong, but since I'm a bit rusty at Java I couldn't find much.

mfietz commented:
> The error description is useless, because the error output/logging is useless. I can't read most of what it says, and the parts I can read (when it complains about wrong credentials) are obviously incorrect.

Still, you being unable to read it (I assume you mean "make any sense of it" here) doesn't mean it won't be any help to us figuring out what is wrong.

But you were able to login to gpodder (with that credentias) at least once? That would make it a server issue.

TwoD commented:
> (I assume you mean "make any sense of it" here) 
No, that's not what I mean. As mentioned earlier, the full text does not fit in the notification area so I can't see anything off screen. All it says is that there was a gpodder sync error, and then it presumably prints a stack trace beginning with "java...", but everything after "java" is cut out.

Yes, as I said, my credentials work on the website (which is back up now), and on the two desktop machines I use the official gPodder client on.

The API appears to be up again and I've spent some time testing things and fixing my Android dev environment. I do get AntennaPod to send sync events now, but I have yet to determine if it responds to actions triggered from other devices.

I'll try keeping an eye on logs and get back when I have something with more substance. Only thing I've got so far is that one of the exceptions was a gpodder.net timeout, but it synced that action again later (but using a timestamp from the second attempt so it appears to have been performed about a minute later than it really did, not really a problem in this case).

Again, sorry for taking up space here, frustration you know...
